File: nQueens.cpp
class Solution {
   
public:
 vector<vector<string>> result;
   vector<vector<bool>> board;
    vector<vector<string>> solveNQueens(int n) {
       this->board[n][n] = {};
       helper(0,n) ;
       return result;
    }
    private:

    void helper(int row, int n)
    {
        //base case

        if(row == n)
        {
            vector<string> temp;
            for(int i = 0; i<n;i++)
            {
                string sb;
                for(int j = 0; j<n;j++)
                {
                    if(board[i][j] )
                    {
                        sb.append("Q");
                    }
                    else{
                        sb.append(".");
                    }
                }
                temp.push_back(sb);
            }

            result.push_back(temp);
            return;
        }

        for(int j = 0;j<n;j++)
        {
            if(isSafe(row,j,n))
            {
                //action
                board[row][j] = true;

                //recurse
                helper(row+1,n);
                //backtrack
                board[row][j] = false;

            }
        }
    }

    bool isSafe(int i,int j,int n)
    {
        int r = i;
        int c = j;

        while(r>=0)
        {
            if(board[r][c])
            return false;
            r--;
        }

        r= i; c = j;
        while(r>=0 && c>=0)
        {
            if(board[r][c])
            return false;
            r--;
            c--;
        }

        r= i; c = j;
        while(r>=0 && c<n)
        {
            if(board[r][c])
            return false;
            r--;
            c--;
        }

        return true;
    }
};

File: N-Queens/N-Queens.java
/* Time Complexity : O(n.n!) 
 *  n - number of queens to be placed */
/* Space Complexity : O(n^2)
 * 	O(n^2) - boolean grid that we prepared */
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this :

//for loop backtrack solution

class Solution {
    List<List<String>> result;
    boolean[][] grid;
    public List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        grid = new boolean[n][n];

        helper(0, n);
        return result;
    }

    private void helper(int r, int n){
        //base condition
        if(r == n){// All the n- queens are placed in the grid
            List<String> li = new ArrayList<>();
            for(int i = 0; i < n; i++){
                StringBuilder sb = new StringBuilder();
                for(int j = 0; j < n; j++){
                    if(grid[i][j]){
                        sb.append('Q');
                    }else{
                        sb.append('.');
                    }
                }
                li.add(sb.toString());                
            }
            result.add(li);
        }
        //logic
        //In a row, check at each cell if it is safe to place the queen
        for(int c = 0; c < n; c++){
            if(isSafe(r,c)){
                //action
                grid[r][c] = true;
                //recurse
                helper(r+1,n);
                //backtrack
                grid[r][c] = false;
            }
        }
    }

    private boolean isSafe(int r, int c){
        //col above
        for(int i = 0; i < r; i++){
            if(grid[i][c]) return false;
        }
        //top left
        int i = r;
        int j = c;
        while(i >= 0 && j >= 0){
            if(grid[i][j]) return false;
            i--;
            j--;
        }
        //top right
        i = r; 
        j = c;
        while(i >= 0 && j < grid.length){
            if(grid[i][j]) return false;
            i--;
            j++;
        }
        return true;
    }
}